src/vpn-api/utils/openvpn.ts

```diff
@@ -0,0 +1,308 @@
+		if (data.substr(0, 5) === '>LOG:') {
+			// >LOG:{timestamp},{level},{message}
+			const logData = data.substr(5);
+			const [, level, ...message] = logData.split(',');
+			this.emit(
+				'log',
+				VpnLogLevels[(level as keyof typeof VpnLogLevels) || 'n'],
+				message.join(','),
+			);
+		} else if (data.substr(0, 8) === '>CLIENT:') {
+			// >CLIENT:{event},{clientId},{...args}
+			// >CLIENT:ENV,{key}={value}
+			// >CLIENT:END
+			for (const eventType of [
+				'connect',
+				'address',
+				'established',
+				'disconnect',
+			]) {
+				if (data.substr(8, eventType.length) === eventType.toUpperCase()) {
+					const [clientId, ...eventArgs] = data
+						.substr(8 + eventType.length + 1)
+						.split(',');
+					this.off('manager:data', this.dataHandler);
+					this.on(
+						'manager:data',
+						this.clientEventEmitterFactory(
+							eventType,
+							parseInt(clientId, 10),
+							eventArgs,
+						),
+					);
+				}
+			}
+		} else if (data.substr(0, 15) === '>BYTECOUNT_CLI:') {
+			// >BYTECOUNT_CLI:{clientId},{bytesRx},{bytesTx}
+			const [clientId, bytesReceived, bytesSent] = data.substr(15).split(',');
+			this.emit('client:bytecount', parseInt(clientId, 10), {
+				bytes_received: parseInt(bytesReceived, 10),
+				bytes_sent: parseInt(bytesSent, 10),
+			});
+		}
+	};
+
+	private clientEventEmitterFactory = (
+		eventName: string,
+		clientId: number,
+		eventArgs?: string[],
+	) => {
+		const env: { [key: string]: number | string } = {};
+		const emitter = (data: string) => {
+			if (data.substr(1, 10) === 'CLIENT:ENV' && data.substr(12) !== 'END') {
+				// >CLIENT:ENV,key=val
+				const msg = data.substr(12);
+				const [key, val] = msg.split('=');
+				env[key] = parsePossibleInt(val);
+			} else {
+				const args: any[] = [clientId];
+				if (eventArgs != null) {
+					for (const arg of eventArgs) {
+						args.push(parsePossibleInt(arg));
+					}
+				}
+				args.push(env);
+				this.off('manager:data', emitter);
+				this.on('manager:data', this.dataHandler);
+				this.emit(`client:${eventName}`, ...args);
+				// re-emit the last message if it was not `CLIENT:ENV,END`
+				if (data.substr(1) !== 'CLIENT:ENV,END') {
+					this.emit('manager:data', data);
+				}
+			}
+		};
+		return emitter;
+	};
+
+	public start(): Bluebird<true> {
+		this.process = spawn('/usr/sbin/openvpn', this.args());
+		// proxy error events from the child process
+		this.process.on('error', err => {
+			this.emit('process:error', err);
+		});
+		this.process.on('exit', (code, signal) => {
+			this.emit('process:exit', code, signal);
+		});
+
+		return this.waitForStart();
+	}
+
+	private waitForStart(since?: number): Bluebird<true> {
+		const now = Date.now();
+		return new Bluebird((resolve, reject) => {
+			if (since != null && now - since > 5000) {
+				return reject(new Error('startup timeout'));
+			}
+			const socket = new net.Socket();
+			const errorHandler = () => {
+				socket.destroy();
+				resolve(this.waitForStart(now));
```

Comment:
I think this would be better as
```ts
reject(new TimeoutError('startup timeout'))
...
reject(new Error('..'))
...
.catch((err) => {
	if (err instanceof TimeoutError) {
		throw err
	}
	return this.waitForStart(now))
}
```
to get back into a standard promise chain as soon as possible rather than having the current unusual pattern